streaming/src/main/scala/org/apache/spark/streaming/receiver/RateLimiter.scala

```diff
@@ -34,12 +36,26 @@ import org.apache.spark.{Logging, SparkConf}
   */
 private[receiver] abstract class RateLimiter(conf: SparkConf) extends Logging {
 
+  // treated as an upper limit
   private val desiredRate = conf.getInt("spark.streaming.receiver.maxRate", 0)
-  private lazy val rateLimiter = GuavaRateLimiter.create(desiredRate)
+  private var currentRate = new AtomicInteger(desiredRate)
+  private lazy val rateLimiter = GuavaRateLimiter.create(currentRate.get())
 
   def waitToPush() {
-    if (desiredRate > 0) {
+    if (currentRate.get() > 0) {
       rateLimiter.acquire()
     }
   }
+
+  private[receiver] def updateRate(newRate: Int): Unit =
+    if (newRate > 0) {
+      try {
+        if (desiredRate > 0) {
+          currentRate.set(newRate min desiredRate)
+        }
+        else currentRate.set(newRate)
```

Comment:
Not sure this is the usual formatting of if expressions in Spark. I would use Java-style, as in the rest of the code base.